### PR TITLE
feature/cplp-1735-delete registration document

### PR DESCRIPTION
**Pull request by qxz2mqe**
https://jira.catena-x.net/browse/CPLP-1735

### CODE REVIEW

File: src/portalbackend/PortalBackend.DBAccess/Repositories/DocumentRepository.cs [hunk 1]
```diff
@@ -95,4 +95,18 @@ public Document CreateDocument(string documentName, byte[] documentContent, byte
...
+    public Task<(Guid DocumentId, DocumentStatusId DocumentStatusId, bool IsSameUser, DocumentTypeId documentTypeId, bool IsApplicationNotSubmitted)> GetDocumentDetailsForApplicationUntrackedAsync(Guid documentId, string iamUserId) =>
+        _dbContext.Documents
+            .AsNoTracking()
+            .Where(x => x.Id == documentId)
+            .Select(document => new ValueTuple<Guid, DocumentStatusId, bool, DocumentTypeId, bool>(
+                document.Id,
+                document.DocumentStatusId,
+                document.CompanyUser!.Company!.CompanyApplications.Any(companyApplication => companyApplication.Company!.CompanyUsers.Any(companyUser => companyUser.IamUser!.UserEntityId == iamUserId)),
+                document.DocumentTypeId,
+                document.CompanyUser!.Company!.CompanyApplications.Any(companyApplication => companyApplication.ApplicationStatusId == CompanyApplicationStatusId.SUBMITTED)
```
Comment (jjeroch) [inline src/portalbackend/PortalBackend.DBAccess/Repositories/DocumentRepository.cs:109]: @qxz2mqe please recheck the logic. As per the requirement the company_application should'nt be in any of the following status 
- submitted
- declined
- confirmed

The code seems to only consider "submitted" and not "declined" and "confirmed"

**Review by jjeroch — DISMISSED** [review]

File: src/registration/Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs [hunk 2]
```diff
@@ -629,4 +629,37 @@ private void UpdateApplicationStatus(CompanyApplication application, UpdateAppli
...
+        {
+            throw new ArgumentException($"Document deletion is not allowed. DocumentType must be either :{string.Join(",", _settings.DocumentTypeIds)}");
+        }
+        if (details.IsApplicationNotSubmitted)
+        {
+            throw new ArgumentException("Document deletion is not allowed. Application is already closed.");
+        }
+        if (details.DocumentId == Guid.Empty)
+        {
+            throw new NotFoundException("Document does not exist.");
```
Comment (jjeroch) [inline src/registration/Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs:646]: @qxz2mqe can you please recheck the exception. I wonder if this exception should not rather be located at the top. If the GUID is empty, the scenarios/exception checks above would not be possible.

File: src/registration/Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs [hunk 3]
```diff
@@ -629,4 +629,37 @@ private void UpdateApplicationStatus(CompanyApplication application, UpdateAppli
...
+        }
+
+        if (!details.IsSameUser)
```
Comment (jjeroch) [inline src/registration/Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs:649]: @qxz2mqe I am a bit unsure about the statement "IsSameUser" - what does this validate?
Important: the user doesn't need to be the same. They only need to belong to the same application.

File: src/portalbackend/PortalBackend.DBAccess/Repositories/DocumentRepository.cs (hunk 1 above)
Comment (qxz2mqe) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/DocumentRepository.cs]: fixed

File: src/registration/Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs (hunk 2 above)
Comment (qxz2mqe) [reply src/registration/Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs]: fixed

File: src/registration/Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs (hunk 3 above)
Comment (qxz2mqe) [reply src/registration/Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs]: IsSameUser rename as IsSameApplicationUser

**Review by ntruchsess — CHANGES_REQUESTED** [review]
please pull main and fix the conflicts so I can to the review

**Comment by qxz2mqe** [comment]
Merged Main and Resolved Conflict

File: src/registration/Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs [hunk 4]
```diff
@@ -610,4 +610,36 @@ private static void UpdateApplicationStatus(CompanyApplication application, Upda
             }
         }
     }
+
+    public async Task<bool> DeleteRegistrationDocumentAsync(Guid documentId, string iamUserId)
+    {
+        var documentRepository = _portalRepositories.GetInstance<IDocumentRepository>();
+        var details = await documentRepository.GetDocumentDetailsForApplicationUntrackedAsync(documentId, iamUserId, _settings.ApplicationStatusIds).ConfigureAwait(false);
+        if (details.DocumentId == Guid.Empty)
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/registration/Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs]: fixed

File: src/registration/Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs [hunk 5]
```diff
@@ -610,4 +610,36 @@ private static void UpdateApplicationStatus(CompanyApplication application, Upda
...
+        {
+            throw new NotFoundException("Document does not exist.");
+        }
+        if (!_settings.DocumentTypeIds.Contains(details.documentTypeId))
+        {
+            throw new ArgumentException($"Document deletion is not allowed. DocumentType must be either :{string.Join(",", _settings.DocumentTypeIds)}");
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/registration/Registration.Service/BusinessLogic/RegistrationBusinessLogic.cs]: fixed

**Review by ntruchsess — APPROVED** [review]

**Pull request merged**